### PR TITLE
feat(patchmix): Add PatchMix augmentation

### LINKED ISSUE #3576: [Feature]: Add PatchMix and TokenMix Augmentations for Vision Transformers
### 🚀 Feature Description

Implement PatchMix and TokenMix augmentations in Kornia’s augmentation mix module. These methods provide patch-level and token-level mixing strategies, which have shown to improve Vision Transformer (ViT) training. The feature will add new classes to kornia.augmentation._2d.mix for both PatchMix and TokenMix, following the API style of existing augmentations like MixUp and CutMix.

### 📂 Feature Category

VLM/VLA Models (Vision Language Models/Agents) - Priority

### 💡 Motivation

Current Kornia augmentations like MixUp and CutMix are widely used, but recent research shows that PatchMix and TokenMix are more effective for Vision Transformers. Adding these will help Kornia users train SOTA ViT models and keep the library up to date with the latest research.

### 💭 Proposed Solution

- Add PatchMix and TokenMix classes to kornia.augmentation._2d.mix.
- Ensure compatibility with Kornia’s API and batching.
- Add unit tests
- 
_Example API:_
 from kornia.augmentation._2d.mix import PatchMix, TokenMix
aug = PatchMix(alpha=1.0, patch_size=8)
out, idx, lam = aug.apply_transform(x, aug.generate_parameters(x.shape), {})

### 🔄 Alternatives Considered

Continue using MixUp or CutMix, but these do not provide the same benefits for ViT architectures as PatchMix and TokenMix.

### 🎯 Use Cases

Use Cases

- Training Vision Transformers on image classification tasks.
- Research on data augmentation for transformer-based vision models.
- Benchmarks comparing augmentation strategies for ViTs.

### 📝 Additional Context

I am interested in implementing this feature for GSoC and will provide a PR with code, tests, and documentation.

PatchMix: https://arxiv.org/abs/2207.08408
tokenMix: https://arxiv.org/abs/2207.08409

### 🤝 Contribution Intent

- [x] I plan to submit a PR to implement this feature
- [ ] I'm requesting this feature but not planning to implement it

### PR DESCRIPTION
**Pull request by Panchadip-128**
⚠️ Issue Link Required: This PR must be linked to an approved and assigned issue. See [Contributing Guide](https://github.com/kornia/kornia/blob/main/CONTRIBUTING.md#pull-request) for details.

Fixes/Relates to: https://github.com/kornia/kornia/issues/3576, Closes https://github.com/kornia/kornia/issues/3576
Closes: #3576

Important:

Ensure you are assigned to the linked issue before submitting this PR
This PR should strictly implement what the linked issue describes
Do not include changes beyond the scope of the linked issue
🛠️ Changes Made
 Added PatchMix export in kornia/augmentation/_2d/__init__.py
 Added PatchMix export in kornia/augmentation/_2d/mix/__init__.py
 Added implementation file: kornia/augmentation/_2d/mix/patchmix.py
 Added/updated tests:
kornia/tests/augmentation/_2d/mix/test_patchmix.py
tests/augmentation/_2d/mix/test_patchmix.py
🧪 How Was This Tested?
 Unit Tests:
python -m pytest -q kornia/tests/augmentation/_2d/mix/test_patchmix.py -q
 Manual Verification:
Verified branch diff scope vs origin/main includes only PatchMix-related files.
 Performance/Edge Cases:
Verified shape/type assertions and gradient flow test in unit tests.
🕵️ AI Usage Disclosure
Check one of the following:

 🟢 No AI used.
 🟡 AI-assisted: I used AI for boilerplate/refactoring but have manually reviewed and tested every line.
 🔴 AI-generated: (Note: These PRs may be subject to stricter scrutiny or immediate closure if the logic is not explained).
🚦 Checklist
 I am assigned to the linked issue (required before PR submission)
 The linked issue has been approved by a maintainer
 I have performed a self-review of my code (no "ghost" variables or hallucinations).
 My code follows the existing style guidelines of this project.
 I have commented my code, particularly in hard-to-understand areas.
 I have added tests that prove my fix is effective or that my feature works.
 (Optional) I have attached screenshots/recordings for UI changes.
💭 Additional Context
This PR is intentionally scoped to one method only (PatchMix) as requested by maintainers.
During local test runs, PyTorch JIT deprecation warnings may appear on Python 3.14+; these warnings are unrelated to this feature change.
image

image

### CODE REVIEW

**Comment by Panchadip-128** [comment]
Hi @Meggison, Please check once, i have implemented all copilot suggestions and other required changes 

Thank you

File: kornia/augmentation/_2d/mix/patchmix.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+from kornia.augmentation import random_generator as rg
+from kornia.augmentation._2d.mix.base import MixAugmentationBaseV2
+from kornia.constants import DataKey
+
+
+class PatchMix(MixAugmentationBaseV2):
+    r"""PatchMix augmentation.
+
+    Replaces a random patch in each image of a batch with the corresponding
+    region from a randomly chosen different image in the batch.
```
Comment (shijianjian) [inline kornia/augmentation/_2d/mix/patchmix.py:33]: Add citation here.

File: kornia/augmentation/_2d/mix/patchmix.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+    Args:
+        alpha: Hyperparameter for the Beta distribution used to generate the
+            mixing parameter lambda.
+        patch_size: The size of the square patch to mix.
+        p: Probability for applying the augmentation.
+        same_on_batch: Apply the same transformation across the batch.
+        keepdim: Whether to keep the output shape the same as input ``True``
+            or broadcast it to the batch form ``False``.
```
Comment (shijianjian) [inline kornia/augmentation/_2d/mix/patchmix.py:45]: add doctests

File: conftest.py [hunk 3]
```diff
@@ -376,6 +297,8 @@ def pytest_report_header(config):
     import kornia_rs
     import onnx
 
+    onnx_version = getattr(onnx, "__version__", "unknown")
```
Comment (shijianjian) [inline conftest.py]: Remove these stuff from this PR.

File: kornia/augmentation/_2d/mix/patchmix.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        out = torch.stack(
+            [
+                input.to(dtype=input.dtype),
+                labels_permute.to(dtype=input.dtype),
+                lam.to(dtype=input.dtype),
+            ],
+            -1,
+        )
+
+        return out
```
Comment (shijianjian) [inline kornia/augmentation/_2d/mix/patchmix.py]: Is this related to a paper? Why design like this?

**Review by shijianjian — COMMENTED** [review]
Please add documentation sample images, following : 

https://github.com/kornia/kornia/blob/main/docs/generate_examples.py

File: kornia/augmentation/_2d/mix/patchmix.py (hunk 2 above)
Comment (Panchadip-128) [reply kornia/augmentation/_2d/mix/patchmix.py:45]: Done - a doctest is now included in the docstring:

python
>>> aug = PatchMix(alpha=1.0, patch_size=4)
>>> x = torch.rand(2, 3, 32, 32)
>>> out = aug(x)
>>> out.shape
torch.Size([2, 3, 32, 32])

File: kornia/augmentation/_2d/mix/patchmix.py (hunk 4 above)
Comment (Panchadip-128) [reply kornia/augmentation/_2d/mix/patchmix.py]: The initial design included `apply_transform_class` and` apply_non_transform_class` to handle label mixing inline, which was an intentional attempt to make the augmentation self-contained. However, after reviewing Kornia's design conventions more carefully, label mixing is meant to be handled at the training loop level outside the augmentation - so I've aligned PatchMix with that pattern in the updated push now.

The `apply_transform` method now performs direct patch replacement on the image tensor only, returning (B, C, H, W). The design follows CutMix-style patch swapping (:yun2019cutmix https://arxiv.org/abs/1905.04899), and the mixing coefficient lam is still available via params["lam"] if the user needs it for soft label computation in their training loop.

**Review by shijianjian — APPROVED** [review]
